backend/src/main/java/fk/prof/backend/model/request/RecordedProfileParser.java

```diff
@@ -0,0 +1,133 @@
+   *
+   * @param codedInputStream
+   * @return starting unread position in buffer
+   */
+  public int parse(CodedInputStream codedInputStream, Buffer underlyingBuffer, int currentPos) throws HttpFailure, AggregationFailure {
+    if (startedAt == null) {
+      startedAt = LocalDateTime.now(Clock.systemUTC());
+    }
+
+    try {
+      if (!headerParser.isParsed()) {
+        currentPos = headerParser.parse(codedInputStream, underlyingBuffer, currentPos);
+
+        if (headerParser.isParsed()) {
+          header = headerParser.get();
+
+          workId = header.getRecordingHeader().getWorkAssignment().getWorkId();
+          aggregationWindow = profileWorkService.getAssociatedAggregationWindow(workId);
+          if (aggregationWindow == null) {
+            throw new HttpFailure(String.format("workId=%d not found, cannot continue receiving associated profile",
+                workId), 400);
+          }
+
+          boolean started = aggregationWindow.startProfile(workId, startedAt);
+          if (!started) {
+            throw new HttpFailure(String.format("Unable to start receiving profile for workId=%d, status=%s",
+                workId, aggregationWindow.getWorkInfo(workId).getStatus()), 400);
+          }
+        }
+      }
+
+      if (headerParser.isParsed()) {
+        while (!codedInputStream.isAtEnd()) {
+          currentPos = wseParser.parse(codedInputStream, underlyingBuffer, currentPos);
+          if (wseParser.isParsed()) {
+            Recorder.Wse wse = wseParser.get();
+            processWse(wse);
+            wseParser.reset();
+          } else {
+            throw new IOException("Incomplete bytes received for wse entry log");
+          }
+        }
+      }
+    } catch (IOException ex) {
+      //NOTE: Ignore this exception. Can happen because incomplete request has been received. Chunks can be received later
+    }
+
+    return currentPos;
+  }
+
+  /**
+   * If parsing was successful, marks the profile as completed/retried, partial otherwise
+   *
+   * @throws HttpFailure
+   */
+  public void close() throws HttpFailure {
+    if (isParsed()) {
+      if (!aggregationWindow.completeProfile(workId)) {
+        throw new HttpFailure(String.format("Unable to complete receiving profile for workId=%d, status=%s",
+            workId, aggregationWindow.getWorkInfo(workId).getStatus()), 400);
+      }
+    } else {
+      if (!aggregationWindow.abandonProfile(workId)) {
+        throw new HttpFailure(String.format("Unable to abandon receiving profile for workId=%d, status=%s",
+            workId, aggregationWindow.getWorkInfo(workId).getStatus()), 400);
+      } else {
+        //Successfully abandoned profile, now raise an exception so that failure http response is returned
+        throw new HttpFailure("Invalid or incomplete payload received", 400);
+      }
+    }
+  }
+
+  private void processWse(Recorder.Wse wse) throws AggregationFailure {
+    indexes.update(wse.getIndexedData());
```

Comment:
chunks are delivered in order and handled by the same event loop thread always. that's one of their selling points.